File: src/main/scala/com/gigahex/samples/analytics/TimeAnalysis.scala
package com.gigahex.samples.analytics

import nl.basjes.parse.useragent.UserAgentAnalyzer
import org.apache.spark.sql.{Row, SparkSession}
import org.apache.spark.sql.functions._

object TimeAnalysis {


  def main(args: Array[String]): Unit = {
    /**
     * Setup the spark session
     */
    val spark = SparkSession.builder()
      .appName("devices_usage")
      .master("local")
      .getOrCreate()
    import spark.implicits._

    val websiteLogs = spark.read.json("hdfs://0.0.0.0:9075/user/gigahex/logs_devices.json")
    val withHour = websiteLogs.withColumn("hour", hour(from_unixtime(col("timestamp")))).cache()
    withHour.show()
    val total = withHour.count()
    withHour.withColumn("range", concat($"hour", lit(" to "), $"hour" + 1))
      .groupBy($"range")
      .count()
      .map(row => (row.getAs[String]("range"),
        row.getAs[Long]("count"),
        (row.getAs[Long]("count") / total.toDouble) * 100))
      .toDF("Time of Day", "Users", "% Users")
      .orderBy(desc("Users"))
      .show()

    spark.stop()


  }

}

File: src/main/py/time_analysis.py
from pyspark.sql import SparkSession
from pyspark.sql.types import Row
from pyspark.sql.functions import *

# Create the spark session

spark = SparkSession.builder.master("local") \
    .appName("count-os-usage") \
    .getOrCreate()

website_logs = spark.read.json("hdfs://0.0.0.0:9075/user/gigahex/logs_devices.json")
with_hour = website_logs.withColumn("hour", hour(to_timestamp(col("timestamp"))))
with_hour.show()
total_users = website_logs.count()
stats = with_hour.withColumn("range", concat(col("hour"), lit(" to "), col("hour") + 1)) \
    .groupBy("range") \
    .count() \
    .rdd \
    .map(lambda row: Row(time_range=row[0],
                         users=row[1],
                         percentage_users=(row[1] / total_users) * 100)).toDF().sort(desc("users"))

stats.show()

spark.stop()
